packages/@ember/object/compat.ts

```diff
@@ -121,23 +122,30 @@ let wrapGetterSetter = function (target: object, key: string, desc: PropertyDesc
   @return {PropertyDecorator} property decorator instance
  */
 export function dependentKeyCompat(
-  target: object,
-  key: string,
-  desc: PropertyDescriptor
+  target: ElementDescriptor[0],
+  key: ElementDescriptor[1],
+  desc: ElementDescriptor[2]
 ): PropertyDescriptor;
-export function dependentKeyCompat(desc: { get?: Function; set?: Function }): Decorator;
-export function dependentKeyCompat(
-  target: object | { get?: Function; set?: Function },
-  key?: string,
-  desc?: PropertyDescriptor
-) {
-  if (!isElementDescriptor([target, key, desc])) {
-    desc = target as PropertyDescriptor;
-
-    let decorator = function (
+export function dependentKeyCompat(desc: PropertyDescriptor): Decorator;
+export function dependentKeyCompat(...args: ElementDescriptor | [PropertyDescriptor]) {
+  if (isElementDescriptor(args)) {
+    let [target, key, desc] = args;
+
+    assert(
+      'The @dependentKeyCompat decorator must be applied to getters/setters when used in native classes',
+      desc != null && (typeof desc.get === 'function' || typeof desc.set === 'function')
+    );
+
+    return wrapGetterSetter(target, key, desc);
+  } else {
+    const desc = args[0];
+
+    assert('expected valid PropertyDescriptor', isPropertyDescriptor(desc));
```

Comment:
Per @chancancode's comment, it might be worth inlining this. It's not in a hot path, but enough of these *do* add up bytes-wise.